File: services/auth-service/src/main/java/com/ba/authservice/service/AuthService.java
package com.ba.authservice.service;

import com.ba.authservice.dto.*;
import com.ba.authservice.model.*;
import com.ba.authservice.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class AuthService {
    
    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final JwtService jwtService;
    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder(12);
    
    @Value("${jwt.refresh-token-expiry}")
    private Long refreshTokenExpiry;
    
    @Value("${security.max-failed-attempts}")
    private Integer maxFailedAttempts;
    
    @Value("${security.lockout-duration}")
    private Integer lockoutDuration;
    
    @Transactional
    public LoginResponse login(LoginRequest request) {
        User user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() -> new RuntimeException("Invalid credentials"));
        
        if (user.isLocked()) {
            throw new RuntimeException("Account is locked. Please try again later.");
        }
        
        if (!user.getActive()) {
            throw new RuntimeException("Account is inactive");
        }
        
        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            user.recordFailedLogin();
            
            if (user.getFailedLoginAttempts() >= maxFailedAttempts) {
                user.lockAccount(lockoutDuration);
            }
            
            userRepository.save(user);
            throw new RuntimeException("Invalid credentials");
        }
        
        user.recordSuccessfulLogin();
        userRepository.save(user);
        
        String accessToken = jwtService.generateAccessToken(user);
        String refreshTokenString = jwtService.generateRefreshToken();
        
        RefreshToken refreshToken = RefreshToken.builder()
                .user(user)
                .token(refreshTokenString)
                .expiresAt(LocalDateTime.now().plusSeconds(refreshTokenExpiry))
                .createdAt(LocalDateTime.now())
                .build();
        
        refreshTokenRepository.save(refreshToken);
        
        return LoginResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshTokenString)
                .tokenType("Bearer")
                .expiresIn(28800L)
                .user(UserDto.builder()
                        .id(user.getId())
                        .username(user.getUsername())
                        .email(user.getEmail())
                        .role(user.getRole())
                        .customerId(user.getCustomerId())
                        .branchId(user.getBranchId())
                        .build())
                .build();
    }
    
    @Transactional
    public void logout(Long userId) {
        refreshTokenRepository.revokeAllUserTokens(userId);
    }
    
    public UserDto getCurrentUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        return UserDto.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .role(user.getRole())
                .customerId(user.getCustomerId())
                .branchId(user.getBranchId())
                .build();
    }
}

File: frontends/customer-portal/src/services/authService.ts
import api from './api';
import { LoginRequest, LoginResponse } from '../types';

interface BackendLoginResponse {
  accessToken: string;
  refreshToken: string;
  tokenType: string;
  expiresIn: number;
  user: {
    id: number;
    username: string;
    email: string;
    role: string;
    customerId?: number;
    branchId?: number;
  };
}

class AuthService {
  async login(credentials: LoginRequest): Promise<LoginResponse> {
    const { data } = await api.post<BackendLoginResponse>('/auth/login', credentials);
    
    // Map backend response to frontend expected format
    const token = data.accessToken;
    const user = {
      ...data.user,
      firstName: data.user.username,
      lastName: '',
    };
    
    localStorage.setItem('authToken', token);
    localStorage.setItem('userData', JSON.stringify(user));
    
    return { token, user };
  }
  
  async logout() {
    await api.post('/auth/logout');
    localStorage.removeItem('authToken');
    localStorage.removeItem('userData');
  }
  
  getCurrentUser() {
    const userData = localStorage.getItem('userData');
    return userData ? JSON.parse(userData) : null;
  }
  
  isAuthenticated() {
    return !!localStorage.getItem('authToken');
  }
}

export default new AuthService();
